### PR TITLE
Add experimental HTTP/3 support

### LINKED ISSUE #537: Allow to specify http3 entrypoints
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you expect to see?

With the addition of http3 in traefik since 2.5.0 it would be nice to be able to specify this feature as a first class option, instead of relying on hacky solutions such as:

>=2.5.0 <2.6.0
```yaml
additionalArguments:
 - "--experimental.http3=true"
 - "--entrypoints.websecure.enabledHTTP3=true"
```

>=2.6.0
```yaml
additionalArguments:
 - "--experimental.http3=true"
 - "--entrypoints.websecure.http3"
 - "--entrypoints.websecure.http3.advertisedPort=443"
```

### PR DESCRIPTION
**Pull request by timoschwarzer**
### What does this PR do?

This PR adds support for experimental HTTP/3 features in Traefik.

### Motivation

HTTP/3 :tada:

Fixes #537

### More

- [X] Yes, I updated the [chart version](https://github.com/traefik/traefik-helm-chart/blob/9b32ed1b414cc0be1ad46bcb335fcfc93ded06f3/traefik/Chart.yaml#L5)

### Additional Notes

I also updated `appVersion` to 2.5.1 because 2.5.0 contained a bug where all `POST` requests failed with HTTP/3.

Co-authored-by: tete17 <miguel_tete17@hotmail.com>

### CODE REVIEW

**Comment by timoschwarzer** [comment]
I labeled this as WIP because it is currently necessary to run Traefik as a privileged container to use port 443. https://github.com/traefik/traefik/issues/8130 needs to be addressed first.

**Comment by kevinpollet** [comment]
Hello @timoschwarzer,

Maybe you have already seen that a [pull request](https://github.com/traefik/traefik-helm-chart/pull/538) has been opened to support HTTP/3 with the `advertisedPort` option introduced in `v2.6-rc1`.

Will you have the time to update your PR or we can close it in favor of https://github.com/traefik/traefik-helm-chart/pull/538?

**Comment by timoschwarzer** [comment]
@kevinpollet thank you for the ping. As far as I can tell right now, #538 is not adding a UDP Kubernetes service for HTTP/3, right? I'll take a look later today.

**Comment by SantoDE** [comment]
The Problem with UDP is, that it will require to be exposed trough a different service of (probably) type `LoadBalancer` as a service can't have multiple protocols in k8s. Im not sure how good the handling of that will be?

**Comment by timoschwarzer** [comment]
@SantoDE We already have two services in total (one for each UDP and TCP). If you enable http/3 for an entrypoint, I add the port to the UDP service [here](https://github.com/traefik/traefik-helm-chart/pull/488/files#diff-61eab19735da47d15ae75df02ed1e41ef123fb5aa1cc1a882e70116b6b58fe0eR6).

**Comment by timoschwarzer** [comment]
@kevinpollet I cherry-picked the work by @tete17, added my changes to the Service manifests and wrote a test for it.

File: traefik/templates/service.yaml [hunk 1]
```diff
@@ -3,9 +3,10 @@
 {{ $tcpPorts := dict }}
 {{ $udpPorts := dict }}
 {{- range $name, $config := .Values.ports }}
-  {{- if eq (toString $config.protocol) "UDP" }}
+  {{- if or $config.http3 (eq (toString $config.protocol) "UDP") }}
     {{ $_ := set $udpPorts $name $config }}
-  {{- else }}
+  {{- end }}
+  {{- if or $config.http3 (eq (toString (default "TCP" $config.protocol)) "TCP") }}
```
Comment (tete17) [inline traefik/templates/service.yaml:9]: Why do we make the extra check here? All http3 endpoints should also contain a tcp port for the backwards http2 compatibility/upgrade-path

File: traefik/tests/service-config_test.yaml [hunk 2]
```diff
@@ -222,3 +222,17 @@ tests:
...
+      ports:
+        websecure:
+          http3:
+            advertisedPort: 8843
+          tls:
+            enabled: true
+    asserts:
+    - equal:
+        path: items[1].spec.ports[0].port
+        value: 8843
```
Comment (tete17) [inline traefik/tests/service-config_test.yaml:238]: Maybe we should also add a check for it to be UDP
```suggestion
    - equal:
        path: items[1].spec.ports[0].port
        value: 8843
    - equal:
        path: items[1].spec.ports[0].protocol
        value: UDP
```

**Review by tete17 — COMMENTED** [review]
Thanks a lot @timoreimann I never thought about changing the load balancer definitions, my changes would not have worked without them.

Looks really good to me. I have given myself the freedom to make a few comments on the final MR that I picked up while reading, mostly nit picking. Feel free to ignore them over the ones that mantainers will do.

I will close my MR in favour of this one :)

File: traefik/templates/service.yaml (hunk 1 above)
Comment (timoschwarzer) [reply traefik/templates/service.yaml:9]: This is what the template does. It adds the port to `udpPorts` if `protocol == 'UDP'`, it adds it to `tcpPorts` if `protocol == 'TCP'` and adds it to both `udpPorts` and `tcpPorts` if `http3 == true`

**Comment by timoschwarzer** [comment]
@kevinpollet I applied a few changes @tete17 suggested. PR should be ready now. :tada:

**Review by kevinpollet — APPROVED** [review]
Hello,

I added @tete17 as a co-author as you worked together on the subject. Just so you know, I updated the code to configure the `advertisedPort` automatically to the exposed port and adapted the tests. Apart from that, awesome work 👍

Thanks again for your contribution.

**Pull request merged**